### PR TITLE
Give project validators ability to block projects

### PR DESCRIPTION
**Pull request by fisher-alice**
Currently, when a verified/authorized teacher reports abuse on a project, the abuse score of the project is increased by `20` so that the project [is blocked](https://github.com/code-dot-org/code-dot-org/blob/672509f2f7373c5c87e78bd92e2170f8b2560f3b/lib/cdo/shared_constants.rb#L83-L86). A project with abuse score `>= 15` is blocked.
https://github.com/code-dot-org/code-dot-org/blob/672509f2f7373c5c87e78bd92e2170f8b2560f3b/apps/src/code-studio/initApp/project.js#L462-L464

However, if the project is frozen, then the project's abuse score cannot be updated because of this line:
https://github.com/code-dot-org/code-dot-org/blob/32f3aac928c84fc22f372ac0eb7f2b88d7024a8a/dashboard/legacy/middleware/helpers/projects.rb#L263

Recently, there was an abuse report via Zendesk on a project that had been an [active featured project so that it was frozen](https://github.com/code-dot-org/code-dot-org/blob/32f3aac928c84fc22f372ac0eb7f2b88d7024a8a/dashboard/app/controllers/featured_projects_controller.rb#L15-L24). Its offending content was not in the App Lab program, but in the data table contents so that the program had to be run a handful of times before seeing offensive text. Customer service has a verified teacher account and so reported abuse from this account. But because it was frozen, the project was unable to be blocked.

Thus, I had to manually unfreeze the project so that I could access and then delete the table. Then I blocked the project. 

In order to make things easier in the future for customer service, this PR allows project validators to report abuse and block projects (by incrementing a project's abuse score by 20) even if the project is frozen. The verified teacher account that customer service uses has both `authorized_teacher` and `project_validator` permissions.

I added a boolean flag parameter to `increment_abuse` in `projects.rb` because this function is called on in `files_api.rb`. If a project's thumbnail image is flagged by image moderation, the abuse score of the project is automatically incremented by 15. Thus, I set the `override_frozen` param to `true` for this call.

https://github.com/code-dot-org/code-dot-org/blob/f39e8b9abba57da9c1ec80d397630fe22a0821ce/dashboard/legacy/middleware/files_api.rb#L1038-L1041

## Links
[Zendesk ticket](https://codeorg.zendesk.com/agent/tickets/516358)

## Testing story
I tested locally using an account with project validator permission and NOT authorized teacher permission and was able to report abuse on projects, even if the project was frozen. A teacher account with only authorized teacher permission still cannot report abuse on frozen projects.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by molly-moen — APPROVED** [review]

File: dashboard/legacy/middleware/helpers/projects.rb [hunk 1]
```diff
@@ -259,8 +259,16 @@ def increment_abuse(channel_id, amount = 10)
 
     row = @table.where(id: project_id).exclude(state: 'deleted').first
     raise NotFound, "channel `#{channel_id}` not found" unless row
+    # If the project is frozen then it is an active featured project or a curriculum exemplar.
```
Comment (bencodeorg) [inline dashboard/legacy/middleware/helpers/projects.rb:262]: Thanks for this comment!

**Review by bencodeorg — APPROVED** [review]
LGTM! One q on this sentence:

>  I will add project validator permission to the verified teacher account that customer service uses.

Are there any side effects of adding project validator permissions to an account (a la "no labs for admins" for admin accounts)?

**Comment by fisher-alice** [comment]
> LGTM! One q on this sentence:
> 
> > I will add project validator permission to the verified teacher account that customer service uses.
> 
> Are there any side effects of adding project validator permissions to an account (a la "no labs for admins" for admin accounts)?

The account customer service uses is not an admin account. It is a teacher account that now has both `authorized_teacher` and `project_validator` permissions.

File: dashboard/legacy/middleware/helpers/projects.rb [hunk 2]
```diff
@@ -254,13 +254,23 @@ def users_paired_on_level?(project_id, current_user_id, owner_user_id, owner_sto
     return true
   end
 
-  def increment_abuse(channel_id, amount = 10)
```
Comment (bencodeorg) [inline dashboard/legacy/middleware/helpers/projects.rb:257]: Just confirming, intentional to remove this default, yeah?

File: dashboard/app/controllers/report_abuse_controller.rb [hunk 3]
```diff
@@ -115,15 +115,15 @@ def update_channel_abuse_score(channel_id)
...
+      if current_user&.verified_teacher? || current_user&.project_validator?
         20
       elsif current_user && !restrict_reporting_to_verified_users
         10
       else
         0
       end
     begin
-      value = Projects.new(get_storage_id).increment_abuse(channel_id, amount)
+      value = Projects.new(get_storage_id).increment_abuse(channel_id, amount, current_user&.project_validator?)
```
Comment (molly-moen) [inline dashboard/app/controllers/report_abuse_controller.rb:127]: nit: mind adding a comment here that project validators can change the abuse score on frozen projects?

File: dashboard/legacy/middleware/files_api.rb [hunk 4]
```diff
@@ -1037,7 +1037,7 @@ def files_put_file(encrypted_channel_id, filename, body)
         case rating
         when :adult, :racy
           # Incrementing abuse score by 15 to differentiate from manually reported projects
-          new_score = project.increment_abuse(encrypted_channel_id, 15)
+          new_score = project.increment_abuse(encrypted_channel_id, 15, true)
```
Comment (molly-moen) [inline dashboard/legacy/middleware/files_api.rb]: nit: also add a comment here that automatic moderation can be applied to frozen projects

**Review by molly-moen — APPROVED** [review]
Looks good! Just a couple comment suggestions for future-proofing 😄

**Review by bencodeorg — APPROVED** [review]

File: dashboard/legacy/middleware/helpers/projects.rb (hunk 2 above)
Comment (fisher-alice) [reply dashboard/legacy/middleware/helpers/projects.rb:257]: Yes - thanks! `increment_abuse` now expects `amount` arg to be passed and `override_frozen` is optional with default value `false`.

**Pull request merged**